File: packages/shared/src/schemas/duo-match.ts
import { z } from "zod";
import { DUO_MATCH_STATES } from "../constants/spoiler";

export const DuoMatchSessionSchema = z.object({
  id: z.string().uuid(),
  initiatorId: z.string().uuid(),
  partnerId: z.string().uuid(),
  state: z.enum(DUO_MATCH_STATES),
  initiatorConsentedAt: z.string().datetime().optional(),
  partnerConsentedAt: z.string().datetime().optional(),
  dataExpiryAt: z.string().datetime().optional(),
  recommendationsGeneratedAt: z.string().datetime().optional(),
  createdAt: z.string().datetime(),
});

export const DuoMatchRecommendationSchema = z.object({
  movie: z.object({
    id: z.string().uuid(),
    title: z.string(),
    posterUrl: z.string().url().optional(),
    genres: z.array(z.string()),
  }),
  jointScore: z.number().min(0).max(10),
  compatibilityReason: z.string(),
  user1PredictedRating: z.number().min(0).max(10),
  user2PredictedRating: z.number().min(0).max(10),
});

export const InitiateDuoMatchSchema = z.object({
  partnerId: z.string().uuid(),
});

export const DuoMatchConsentSchema = z.object({
  consent: z.boolean(),
});

export type DuoMatchSession = z.infer<typeof DuoMatchSessionSchema>;
export type DuoMatchRecommendation = z.infer<typeof DuoMatchRecommendationSchema>;
export type InitiateDuoMatch = z.infer<typeof InitiateDuoMatchSchema>;
export type DuoMatchConsent = z.infer<typeof DuoMatchConsentSchema>;

File: services/api/app/api/v1/schemas/duo_match.py
from pydantic import BaseModel


class InitiateDuoMatchRequest(BaseModel):
    partner_id: str


class DuoMatchConsentRequest(BaseModel):
    consent: bool


class DuoMatchSessionResponse(BaseModel):
    id: str
    initiator_id: str
    partner_id: str
    state: str
    initiator_consented_at: str | None = None
    partner_consented_at: str | None = None
    data_expiry_at: str | None = None
    recommendations_generated_at: str | None = None
    created_at: str

    model_config = {"from_attributes": True}


class DuoMatchRecommendationResponse(BaseModel):
    movie: dict  # {id, title, posterUrl, genres}
    joint_score: float
    compatibility_reason: str
    user1_predicted_rating: float
    user2_predicted_rating: float
